File: python/pr2_sim/arm.py
import numpy as np

import roslib
roslib.load_manifest('tfx')
import tfx
import tf.transformations as tft

import openravepy as rave

import simulator
from utils import utils

import IPython

class Arm:
    """
    Class for controlling the end effectors of the PR2 in OpenRAVE
    """
    
    joint_name_suffixes = ["_shoulder_pan_joint",
                           "_shoulder_lift_joint",
                           "_upper_arm_roll_joint",
                           "_elbow_flex_joint",
                           "_forearm_roll_joint",
                           "_wrist_flex_joint",
                           "_wrist_roll_joint"]
    
    L_POSTURES = {
                  'untucked' : [0.4,  1.0,    0.0,  -2.05,  0.0,  -0.1,   0.0],
                  'tucked'   : [0.06, 1.25,   1.79, -1.68, -1.73, -0.10, -0.09],
                  'up'       : [0.33, -0.35,  2.59, -0.15,  0.59, -1.41, -0.27],
                  'side'     : [1.83, -0.33,  1.01, -1.43,  1.1,  -2.10,  3.07],
                  'mantis'   : [2.03, -0.054, 1.01, -1.47,  0.55, -1.42,  0.82]
                  }
    
    
    def __init__(self, arm_name, sim=None, view=True):
        """
        :param arm_name: "left" or "right"
        :param sim: OpenRave simulator (or create if None)
        """
        assert arm_name == 'left' or arm_name == 'right'
        
        self.arm_name = arm_name
        self.tool_frame = '{0}_gripper_tool_frame'.format(arm_name[0])
        self.joint_names = ['{0}{1}'.format(arm_name[0], joint_name) for joint_name in self.joint_name_suffixes]
        
        self.sim = sim
        if self.sim is None:
            self.sim = simulator.Simulator()
              
            if view:
                self.sim.env.SetViewer('qtcoin')
            
        self.robot = self.sim.robot
        self.manip = self.sim.larm if arm_name == 'left' else self.sim.rarm   
        self.tool_frame_link = self.robot.GetLink(self.tool_frame)
        
        self.gripper_joint = self.robot.GetJoint('{0}_gripper_l_finger_joint'.format(arm_name[0]))
        
        self.joint_indices = self.manip.GetArmIndices()

    #####################
    # command methods   #
    #####################
    
    def set_posture(self, posture):
        """
        :param posture: 'untucked', 'tucked', 'up', 'side', 'mantis'
        """
        assert posture in self.L_POSTURES.keys()
        
        l_joints = self.L_POSTURES[posture]
        joints = l_joints if self.arm_name == "left" else Arm._mirror_arm_joints(l_joints)
        self.set_joints(joints)
    
    def set_pose(self, pose):
        """
        :param pose: tfx.pose
        """
        joints = self.ik(pose)
        if joints is not None:
            self.set_joints(joints)
    
    def set_joints(self, joint_values):
        """
        :param joints: list of joint values
        """
        l_limits, u_limits = self.get_joint_limits()
        valid_joint_values = np.array(joint_values)
        valid_joint_values = np.maximum(joint_values, l_limits)
        valid_joint_values = np.minimum(joint_values, u_limits)
        if not np.min(np.array(joint_values) == valid_joint_values):
            print('Invalid joints. Setting invalid values to limits')
        self.robot.SetDOFValues(joint_values, self.joint_indices)
        
    def teleop(self):
        print('{0} arm teleop'.format(self.arm_name))
        
        pos_step = .01
        delta_position = {'a' : [0, pos_step, 0],
                          'd' : [0, -pos_step, 0],
                          'w' : [pos_step, 0, 0],
                          'x' : [-pos_step, 0, 0],
                          '+' : [0, 0, pos_step],
                          '-' : [0, 0, -pos_step]}
        
        angle_step = 2.0
        delta_angle = {'o' : [angle_step, 0, 0],
                       'p' : [-angle_step, 0, 0],
                       'k' : [0, angle_step, 0],
                       'l' : [0, -angle_step, 0],
                       'n' : [0, 0, angle_step],
                       'm' : [0, 0, -angle_step]}
        
        char = ''
        while char != 'q':
            char = utils.Getch.getch()
            pose = self.get_pose()
            new_pose = tfx.pose(pose)
            if delta_position.has_key(char):
                new_pose.position = pose.position.array + delta_position[char]
            ypr = np.array([pose.tb_angles.yaw_deg, pose.tb_angles.pitch_deg, pose.tb_angles.roll_deg])
            if delta_angle.has_key(char):
                ypr += np.array(delta_angle[char])
                new_pose = tfx.pose(pose.position, tfx.tb_angles(ypr[0], ypr[1], ypr[2]))
            new_joints = self.ik(new_pose)
            if new_joints is not None:
                self.set_joints(new_joints)
            else:
                print('teleop: IK invalid')
            
        print('{0} arm end teleop'.format(self.arm_name))
        
    def open_gripper(self):
        self.set_gripper(1.0)
    
    def close_gripper(self):
        self.set_gripper(0.0)
    
    def set_gripper(self, pct):
        """
        Set gripper to percentage open
        
        :param pct: [0,1] amount the gripper is open
        """
        lower, upper = self.gripper_joint.GetLimits()
        value = pct*(upper[0] - lower[0]) + lower[0]
        index = self.gripper_joint.GetDOFIndex()
        self.robot.SetDOFValues([value], [index])
        
    
    #######################
    # state info methods  #
    #######################
    
    def get_pose(self):
        """
        :return current pose of tool_frame as tfx.pose
        """
        return tfx.pose(self.sim.transform_from_to(np.eye(4), self.tool_frame, 'base_link'), frame='base_link')
    
    def get_joints(self):
        """
        :return current joint values
        """
        return self.manip.GetArmDOFValues()
    
    def get_joint_limits(self):
        """
        :return [lower limits array, upper limits array]
        """
        l_limits, u_limits = self.robot.GetDOFLimits()
        arm_l_limits = np.array([l_limits[i] for i in self.joint_indices])
        arm_u_limits = np.array([u_limits[i] for i in self.joint_indices])
        return arm_l_limits, arm_u_limits
    
    
    #############
    #   fk/ik   #
    #############
    
    def fk(self, joints):
        """
        :return tfx.pose of tool_frame pose
        """
        curr_joints = self.get_joints()
        
        self.set_joints(joints)
        pose = self.get_pose()
    
        self.set_joints(curr_joints)
        return pose
    
    def ik(self, pose):
        """
        :param pose: tfx.pose
        :return list of joints or None if no solution
        """
        assert pose.frame == 'base_link'
        
        pose_mat_world = self.sim.transform_from_to(pose.matrix, pose.frame, 'world')
        joints = self.sim.ik_for_link(pose_mat_world, self.manip, self.tool_frame, 0)
        
        if joints is not None:
            joints = self._closer_joint_angles(joints, self.get_joints())
            
        return joints
    
    def ik_lookat(self, position, point):
        """
        :param position: desired position of tool_frame (tfx.point)
        :param point: desired point for tool_frame to point at (tfx.point)
        """
        assert position.frame == 'base_link'
        
        position_world = self.sim.transform_from_to(position.array, 'base_link', 'world')
        direction_world = np.dot(self.sim.transform_from_to(np.eye(4), 'base_link', 'world')[:3,:3], point.array - position.array)
        direction_world /= np.linalg.norm(direction_world)
        
        red = direction_world
        red /= np.linalg.norm(red)
        green = np.cross(red, np.array([0,0,1]))
        green /= np.linalg.norm(red)
        blue = np.cross(red, green)
        blue /= np.linalg.norm(blue)
        
        pose_mat = np.eye(4)
        pose_mat[:3,0] = red
        pose_mat[:3,1] = green
        pose_mat[:3,2] = blue
        pose_mat[:3,3] = position_world
        
        pose = tfx.pose(self.sim.transform_from_to(pose_mat, 'world', 'base_link'), frame='base_link')
        return self.ik(pose)
        
    ##################
    # helper methods #
    ##################
    
    @staticmethod
    def _mirror_arm_joints(x):
        """"
        Mirror image of joints (r->l or l->r)
        
        :param x: joints
        """
        return np.array([-x[0],x[1],-x[2],x[3],-x[4],x[5],-x[6]])
    
    @staticmethod
    def _closer_joint_angles(new_joints, curr_joints):
        for i in [2, 4, 6]:
            new_joints[i] = utils.closer_angle(new_joints[i], curr_joints[i])
        return new_joints
    
#############
#   TESTS   #
#############
    
def test_ik():
    arm = Arm('right')
    arm.set_posture('mantis')
    
    actual_joints = arm.get_joints()
    actual_pose = arm.get_pose()
    
    ik_joints = arm.ik(actual_pose)
    
    print('actual_joints: {0}'.format(actual_joints))
    print('ik_joints: {0}'.format(ik_joints))

def test_teleop():
    arm = Arm('right')
    arm.set_posture('mantis')
    
    arm.teleop()
    
    
if __name__ == '__main__':
    test_ik()
    #test_teleop()

File: src/pr2_utils/pr2_sim/arm.cpp
#include "pr2_utils/pr2_sim/arm.h"

namespace pr2_sim {

/**
 *
 * Arm constructors
 *
 */

Arm::Arm(ArmType a, Simulator *s) : arm_type(a), sim(s) {
	if (s == NULL) {
		sim = new Simulator();
	}

	tool_frame = (arm_type == ArmType::left) ? "l_gripper_tool_frame" : "r_gripper_tool_frame";
	manip_name = (arm_type == ArmType::left) ? "leftarm" : "rightarm";
	manip = sim->robot->GetManipulator(manip_name);

	joint_indices = manip->GetArmIndices();
	num_joints = joint_indices.size();

	std::vector<double> lower_vec(num_joints), upper_vec(num_joints);
	sim->robot->GetDOFLimits(lower_vec, upper_vec, joint_indices);

	lower = VectorJ(lower_vec.data());
	upper = VectorJ(upper_vec.data());

	lower(5) = -M_PI/2; // enforce _wrist_flex_joint so carmine doesn't collide
	upper(5) = M_PI/2;

	sim->update(); // must do so the origin transform is correct
	Matrix4d eye = Matrix4d::Identity();
	fk_origin = eigen_to_rave(sim->transform_from_to(eye, "torso_lift_link", "base_link"));

	arm_joint_axes = { rave::Vector(0,0,1),
			rave::Vector(0,1,0),
			rave::Vector(1,0,0),
			rave::Vector(0,1,0),
			rave::Vector(1,0,0),
			rave::Vector(0,1,0),
			rave::Vector(1,0,0) };

	arm_link_trans = { rave::Vector(0, (arm_type == ArmType::left) ? 0.188 : -0.188, 0),
			rave::Vector(0.1, 0, 0),
			rave::Vector(0, 0, 0),
			rave::Vector(0.4, 0, 0),
			rave::Vector(0, 0, 0),
			rave::Vector(.321, 0, 0),
			rave::Vector(.18, 0, 0) };
}

/**
 * Arm command methods
 */

void Arm::set_posture(Arm::Posture posture) {
	std::vector<double> j;

	// joint values defined for left arm
	switch(posture) {
	case Posture::untucked:
		j = {0.4,  1.0,   0.0,  -2.05,  0.0,  -0.1,  0.0};
		break;
	case Posture::tucked:
		j = {0.06, 1.25, 1.79, -1.68, -1.73, -0.10, -0.09};
		break;
	case Posture::up:
		j = {0.33, -0.35,  2.59, -0.15,  0.59, -1.41, -0.27};
		break;
	case Posture::side:
		j = {1.832,  -0.332,   1.011,  -1.437,   1.1  ,  -2.106,  3.074};
		break;
	case Posture::mantis:
		j = {2.03018192, -0.05474993, 1.011, -1.47618716,  0.55995636, -1.42855926,  0.823};
		break;
	default:
		j = {0, 0, 0, 0, 0, 0, 0};
		break;
	}

	if (arm_type == ArmType::right) {
		j = {-j[0], j[1], -j[2], j[3], -j[4], j[5], -j[6]};
	}

	VectorJ j_vec(j.data());
	set_joints(j_vec);
}

void Arm::set_pose(const Matrix4d& pose) {
	VectorJ joints;
	if (ik(pose, joints)) {
		set_joints(joints);
	} else {
		std::cout << "Cannot set arm pose to:\n" << pose << "\nIK failed\n";
	}
}

void Arm::set_joints(const VectorJ& joint_values) {
	std::vector<double> joint_values_vec(joint_values.rows());
	for(int i=0; i < num_joints; ++i) {
		joint_values_vec[i] = std::min(joint_values(i), upper(i));
		joint_values_vec[i] = std::max(joint_values(i), lower(i));
	}

	sim->robot->SetDOFValues(joint_values_vec, rave::KinBody::CheckLimitsAction::CLA_Nothing, joint_indices);
}

void Arm::teleop() {
	double pos_step = .01;
	std::map<int,rave::Vector> delta_position =
	{
			{'a' , rave::Vector(0, pos_step, 0)},
			{'d' , rave::Vector(0, -pos_step, 0)},
			{'w' , rave::Vector(pos_step, 0, 0)},
			{'x' , rave::Vector(-pos_step, 0, 0)},
			{'+' , rave::Vector(0, 0, pos_step)},
			{'-' , rave::Vector(0, 0, -pos_step)},
	};

	double angle_step = 2.0*(M_PI/180);
	std::map<int,rave::Vector> delta_angle =
	{
			{'p', rave::Vector(angle_step, 0, 0)},
			{'o', rave::Vector(-angle_step, 0, 0)},
			{'k', rave::Vector(0, angle_step, 0)},
			{'l', rave::Vector(0, -angle_step, 0)},
			{'n', rave::Vector(0, 0, angle_step)},
			{'m', rave::Vector(0, 0, -angle_step)},
	};

	std::cout << manip_name << " teleop\n";

	char c;
	while ((c = pr2_utils::getch()) != 'q') {

		rave::Transform pose = eigen_to_rave(get_pose());
		if (delta_position.count(c) > 0) {
			pose.trans += delta_position[c];
		} else if (delta_angle.count(c) > 0) {
			pose.rot = rave::geometry::quatFromAxisAngle(rave::geometry::axisAngleFromQuat(pose.rot) + delta_angle[c]);
		}
		set_pose(rave_to_eigen(pose));
	}

	std::cout << manip_name << " end teleop\n";
}

/**
 * Arm state info methods
 */

Matrix4d Arm::get_pose() {
	Matrix4d pose = Matrix4d::Identity();
	return sim->transform_from_to(pose, tool_frame, "base_link");
}

VectorJ Arm::get_joints() {
	std::vector<double> joint_values;
	sim->robot->GetDOFValues(joint_values, joint_indices);
	return VectorJ(joint_values.data());
}

void Arm::get_joint_limits(VectorJ& l, VectorJ& u) {
	l = lower;
	u = upper;
}

/**
 * Arm fk/ik
 */

Matrix4d Arm::fk(const VectorJ& joints) {
	rave::Transform pose_mat = fk_origin;

	rave::Transform R;
	for(int i=0; i < ARM_DIM; ++i) {
		R.rot = rave::geometry::quatFromAxisAngle(arm_joint_axes[i], joints(i));
		R.trans = arm_link_trans[i];
		pose_mat = pose_mat*R;
	}

	return rave_to_eigen(pose_mat);

//	VectorJ curr_joints = get_joints();
//
//	set_joints(joints);
//	Matrix4d pose = get_pose();
//
//	set_joints(curr_joints);
//	return pose;
}


/**
 * \param pose desired pose of tool_frame in frame "base_link"
 */
bool Arm::ik(const Matrix4d& pose, VectorJ& joints) {
	Matrix4d pose_world = sim->transform_from_to(pose, "base_link", "world");

	std::vector<double> joint_vec;
	if (sim->ik_for_link(pose_world, manip, tool_frame, joint_vec)) {
		joints = VectorJ(joint_vec.data()); // TODO: add closer_angle
		sim->update();
//		joints = closer_joint_angles(joints, get_joints());
		return true;
	} else {
		return false;
	}
}

/**
 * \brief Finds a rotation matrix that looks at point from position, then calls usual IK
 *        (will not always find a solution, even when it should)
 * \param position desired gripper position in frame "base_link"
 * \param point 3d point to look at in frame "base_link"
 */
bool Arm::ik_lookat(const Vector3d& position, const Vector3d& point, VectorJ& joints) {
	Vector3d position_world = sim->transform_from_to(position, "base_link", "world");
	Matrix4d eye = Matrix4d::Identity();
	Vector3d direction_world = sim->transform_from_to(eye, "base_link", "world").block<3,3>(0,0)*(point - position);
	direction_world.normalize();

	Vector3d red, green, blue;
	red = direction_world;
	red.normalize();
	green = red.cross(Vector3d(0,0,1));
	green.normalize();
	blue = red.cross(green);
	blue.normalize();

	Matrix4d pose_world = Matrix4d::Identity();
	pose_world.block<3,3>(0,0) << red, green, blue;
	pose_world.block<3,1>(0,3) = position_world;

	Matrix4d pose_base = sim->transform_from_to(pose_world, "world", "base_link");
	return ik(pose_base, joints);
}

/**
 *
 * Arm private methods
 *
 */



}
